### PR TITLE
Color phylo tree viz by sample name

### PR DESCRIPTION
**Pull request by jshoe**
- Color the tree by a given attribute name recursively
- Change previous 'highlight' behavior to bold the path to the nodes
- Pass down projects.name in phylo_tree.rb
- Color in this photo is a little bad but Colormap will be improved

### CODE REVIEW

File: app/assets/src/styles/views/phylo_tree/phylo_tree_vis.scss [hunk 1]
```diff
@@ -12,14 +12,9 @@
   }
 
   .node circle {
-    fill: $light-grey;
     opacity: 0.8;
```
Comment (tfrcarvalho) [inline app/assets/src/styles/views/phylo_tree/phylo_tree_vis.scss]: Can you keep it grey here? This is the default color.
Currently, we do not have a color if we disable the highlight per attribute.

Also, do you know which takes precedence fill in JS or in CSS?

File: app/assets/src/styles/views/phylo_tree/phylo_tree_vis.scss [hunk 2]
```diff
@@ -12,14 +12,9 @@
...
   }
 
-  .node.highlight circle {
-    fill: $primary-light;
```
Comment (tfrcarvalho) [inline app/assets/src/styles/views/phylo_tree/phylo_tree_vis.scss]: Same here.

File: app/assets/src/styles/views/phylo_tree/phylo_tree_vis.scss [hunk 3]
```diff
@@ -28,22 +23,18 @@
     font-weight: 700;
   }
 
-  .node-internal circle {
-    fill: #ccc;
```
Comment (tfrcarvalho) [inline app/assets/src/styles/views/phylo_tree/phylo_tree_vis.scss]: Same here. Maybe change it to the closes color in the `_colors.scss`

File: app/assets/src/components/visualizations/dendrogram/Dendogram.js [hunk 4]
```diff
@@ -147,6 +149,110 @@ export default class PhyloTree {
...
+  }
+
+  updateLegend() {
+    // Generate legend for coloring by attribute name
+    let other = this.options.colorGroupMissingName || "Other";
+    this._attrValToColor[other] = this.options.absentColor || "#000000";
+    let legend = this.g.select(".legend");
+    if (legend.empty()) {
+      legend = this.g.append("g").attr("class", "legend");
+      let x = 900;
```
Comment (tfrcarvalho) [inline app/assets/src/components/visualizations/dendrogram/Dendogram.js]: Can you move x and y to options?

File: app/assets/src/components/visualizations/dendrogram/Dendogram.js [hunk 5]
```diff
@@ -147,6 +149,110 @@ export default class PhyloTree {
...
+      return colorResult;
+    }
+
+    colorThisNode(this.root, this._attrValToColor);
+  }
+
+  updateLegend() {
+    // Generate legend for coloring by attribute name
+    let other = this.options.colorGroupMissingName || "Other";
+    this._attrValToColor[other] = this.options.absentColor || "#000000";
```
Comment (tfrcarvalho) [inline app/assets/src/components/visualizations/dendrogram/Dendogram.js]: Again, do not redefine absent color.

File: app/assets/src/components/visualizations/dendrogram/Dendogram.js [hunk 6]
```diff
@@ -147,6 +149,110 @@ export default class PhyloTree {
...
+        .append("text")
+        .attr("x", x)
+        .attr("y", y - 20)
+        .text(legendTitle);
+
+      for (const attrVal in this._attrValToColor) {
+        // Add color box
+        let color = this._attrValToColor[attrVal];
+        legend
+          .append("rect")
```
Comment (tfrcarvalho) [inline app/assets/src/components/visualizations/dendrogram/Dendogram.js]: I believe Jenn's prototype was a circle here.

File: app/assets/src/components/visualizations/dendrogram/Dendogram.js [hunk 7]
```diff
@@ -345,6 +451,8 @@ export default class PhyloTree {
     let maxDistance = this.computeDistanceToRoot(this.root);
     this.createScale(-30, 0, this.minTreeSize.width, maxDistance);
     this.updateHighlights();
+    this.updateColors();
```
Comment (tfrcarvalho) [inline app/assets/src/components/visualizations/dendrogram/Dendogram.js]: This does not need to run for all update calls. Only when the tree is first drawn or rerooted.

**Comment by jshoe** [comment]
thanks @tfrcarvalho, will work on this

**Comment by jshoe** [comment]
When you don't pass in options:
screen shot 2018-09-25 at 2 37 25 pm

New look:
screen shot 2018-09-25 at 2 37 51 pm

Jenn's previous comments:
- I would make the circles smaller and opaque
- Also I’d make the text all black instead of color
- it would be great if you can add a slight rounded corner on the tree
- i’d also make Project Name bolded and move the names closer to the square, like the same distance the sample names are from the circles
- the 3 colors should be another combo, ill suggest something

File: app/assets/src/components/visualizations/dendrogram/Dendogram.js [hunk 8]
```diff
@@ -147,6 +150,113 @@ export default class PhyloTree {
...
+    });
+    allVals = Array.from(allVals);
+    allVals = ["Uncolored"]
+      .concat(allVals)
+      .concat(this.options.colorGroupAbsentName);
+    this.allColorAttributeValues = allVals;
+
+    // Set up colors array
+    this.colors = Colormap.getNScale(this.options.colormapName, allVals.length);
+    this.colors = ["#cccccc"].concat(this.colors);
```
Comment (tfrcarvalho) [inline app/assets/src/components/visualizations/dendrogram/Dendogram.js]: Can you move the color to an option?
I just do not link having magic values in the code.

Comment (jshoe) [reply app/assets/src/components/visualizations/dendrogram/Dendogram.js]: 👍

**Pull request merged**